source/concepts.tex

```diff
@@ -0,0 +1,1249 @@
+\tcode{is_swappable_v<T>}.
+\end{itemdescr}
+
+\indexlibrary{\idxcode{SwappableWith}}%
+\begin{itemdecl}
+template <class T, class U>
+concept SwappableWith =
+  is_swappable_with_v<T, T> && is_swappable_with_v<U, U> &&
+  CommonReference<const remove_reference_t<T>&, const remove_reference_t<U>&> &&
+  is_swappable_with_v<T, U> && is_swappable_with_v<U, T>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+Let:
+\begin{itemize}
+\item \tcode{t1} and \tcode{t2} denote distinct equal objects of type
+  \tcode{remove_cvref_t<T>},
+\item $E_t$ be an expression that denotes \tcode{t1} such that
+  \tcode{decltype(($E_t$))} is \tcode{T},
+\item \tcode{u1} and \tcode{u2} similarly denote distinct equal objects of type
+  \tcode{remove_cvref_t<U>},
+\item $E_u$ be an expression that denotes \tcode{u1} such that
+  \tcode{decltype(($E_u$))} is \tcode{U}, and
+\item \tcode{C} be
+  \begin{codeblock}
+    common_reference_t<const remove_reference_t<T>&, const remove_reference_t<U>&>
+  \end{codeblock}
+\end{itemize}
+\tcode{SwappableWith<T, U>} is satisfied only if after evaluating either
+\tcode{swap($E_t$, $E_u$)} or \tcode{swap($E_u$, $E_t$)} in the context described
+above, \tcode{C(t1)} is equal to \tcode{C(u2)} and \tcode{C(u1)} is equal to \tcode{C(t2)}.
+
+\pnum
+The context in which \tcode{swap($E_t$, $E_u$)} or \tcode{swap($E_u$, $E_t$)} are evaluated shall ensure that a binary non-member
+function named "swap" is selected via overload resolution\iref{over.match} on a candidate set that includes:
+\begin{itemize}
+\item the two \tcode{swap} function templates defined in \tcode{<utility>}~(\ref{utility}) and
+\item the lookup set produced by argument-dependent lookup\iref{basic.lookup.argdep}.
+\end{itemize}
+
+\pnum
+There need be no subsumption relationship between \tcode{SwappableWith<T, U>} and
+any specialization of \tcode{is_swappable_with_v}.
+\end{itemdescr}
+
+\begin{itemdescr}
+\pnum
+This subclause provides definitions for swappable types and expressions. In these
+definitions, let \tcode{t} denote an expression of type \tcode{T}, and let \tcode{u}
+denote an expression of type \tcode{U}.
+
+\pnum
+An object \tcode{t} is \defn{swappable with} an object \tcode{u} if and only if
+\tcode{SwappableWith<T, U>} is satisfied. \tcode{Swappable\-With<T, U>} is satisfied
+only if given distinct objects \tcode{t2} equal to \tcode{t}
+and \tcode{u2} equal to \tcode{u}, after evaluating either
+\tcode{ranges::swap(t, u)} or \tcode{ranges::swap(u, t)}, \tcode{t2} is equal to
+\tcode{u} and \tcode{u2} is equal to \tcode{t}.
+
+\pnum
+An rvalue or lvalue \tcode{t} is \defn{swappable} if and only if \tcode{t} is
+swappable with any rvalue or lvalue, respectively, of type \tcode{T}.
+
+\begin{example}
+User code can ensure that the evaluation of \tcode{swap} calls
+is performed in an appropriate context under the various conditions as follows:
+\begin{codeblock}
+#include <utility>
+
+template <class T, SwappableWith<T> U>
```

Comment:
and a `std::`.

Reply:
or two.

Reply:
How so? We usually don't prefix with `std::` in the library sections of the standard, but maybe I'm overlooking something.

Reply:
We also don't usually include headers in library examples. Since this particular example is hypothetical user code, not library code, with a `main()` and all, I think it makes more sense to write it like a user would.

Reply:
Elaborating @timsong-cpp's statement: I dropped this example in R2 of the proposal, and Daniel Krügler specifically asked me to replace it. This particular example is crucial to readers understanding how the ADL 2-step works in practice. It needs full qualification to make it clear how the namespaces and qualifications interact.